### PR TITLE
Gpl 770 beckman cherrypicked samples

### PR DESCRIPTION
**Pull request by Chris-Friend**
* Add to the `get_cherrypicked_samples` method a query to identify samples that have been cherrypicked by the Beckman workflow.
* Also lots of test to check the complexities

### CODE REVIEW

File: tests/helpers/test_reports_helpers.py [hunk 1]
```diff
@@ -74,35 +83,42 @@ def test_delete_reports(app, freezer):
...
-    # Note: This represents the results of three different database queries
-    # each of which gets indexed from 0. Do not changes the indicies here unless
-    # you have modified the behaviour of the query.
+def test_get_cherrypicked_samples_chunking_no_beckman(app, freezer):
+    # Note: This represents the results of three different (Sentinel, Beckman) sets of
+    # database queries, each Sentinel query getting indexed from 0. Do not changes the
+    # indicies here unless you have modified the behaviour of the query.
     query_results = [
         pd.DataFrame(["MCM001"], columns=[FIELD_ROOT_SAMPLE_ID], index=[0]),
+        pd.DataFrame([]),
```
Comment (andrewsparkes) [inline tests/helpers/test_reports_helpers.py:117]: maybe add the beckman comment for these empty dataframes so we know why they are here like above

File: tests/helpers/test_reports_helpers.py [hunk 2]
```diff
@@ -144,6 +160,170 @@ def test_get_cherrypicked_samples_repeat_tests(
...
+                assert returned_samples.at[1, FIELD_ROOT_SAMPLE_ID] == "MCM003"
+                assert returned_samples.at[2, FIELD_ROOT_SAMPLE_ID] == "MCM005"
+
+
+def test_get_cherrypicked_samples_chunking_no_sentinel(app, freezer):
+    # Note: This represents the results of three different (Sentinel, Beckman) sets of
+    # database queries, each Sentinel query getting indexed from 0. Do not changes the
+    # indicies here unless you have modified the behaviour of the query.
+    query_results = [
+        pd.DataFrame([]),
```
Comment (andrewsparkes) [inline tests/helpers/test_reports_helpers.py:190]: and maybe sentinel and beckman comments here too

File: tests/helpers/test_reports_helpers.py [hunk 3]
```diff
@@ -144,6 +160,170 @@ def test_get_cherrypicked_samples_repeat_tests(
...
+# test scenario where there have been multiple lighthouse tests for a sample with the same Root
+# Sample ID uses actual databases rather than mocking to make sure the query is correct
+def test_get_cherrypicked_samples_repeat_tests_no_sentinel(
+    app, freezer, mlwh_beckman_cherrypicked, event_wh_data
+):
+    # the following come from MLWH_SAMPLE_LIGHTHOUSE_SAMPLE in fixture_data
+    root_sample_ids = ["root_4", "root_5", "root_4"]
+    plate_barcodes = ["pb_4", "pb_5", "pb_6"]
+
+    # root_4 will match 2 samples, but only one of those will match an event (on sample uuid)
```
Comment (andrewsparkes) [inline tests/helpers/test_reports_helpers.py:223]: yep, good, i like the comments clearly describing why we get the result in your tests

File: tests/helpers/test_reports_helpers.py [hunk 4]
```diff
@@ -144,6 +160,170 @@ def test_get_cherrypicked_samples_repeat_tests(
...
+                returned_samples = get_cherrypicked_samples(samples, plate_barcodes)
+                assert returned_samples.at[0, FIELD_ROOT_SAMPLE_ID] == "MCM001"
+                assert returned_samples.at[1, FIELD_ROOT_SAMPLE_ID] == "MCM006"
+                assert returned_samples.at[2, FIELD_ROOT_SAMPLE_ID] == "MCM003"
+                assert returned_samples.at[3, FIELD_ROOT_SAMPLE_ID] == "MCM005"
+
+
+def test_get_cherrypicked_samples_chunking_sentinel_and_beckman(app, freezer):
+    # Note: This represents the results of three different (Sentinel, Beckman) sets of
+    # database queries, each query getting indexed from 0. Do not changes the
```
Comment (andrewsparkes) [inline tests/helpers/test_reports_helpers.py:264]: again typos changes -> change and indicies -> indices

File: tests/helpers/test_reports_helpers.py [hunk 5]
```diff
@@ -144,6 +160,170 @@ def test_get_cherrypicked_samples_repeat_tests(
...
+    expected_columns = [FIELD_ROOT_SAMPLE_ID, FIELD_PLATE_BARCODE, "Result_lower", FIELD_COORDINATE]
+    expected = pd.DataFrame(np.array(expected_rows), columns=expected_columns, index=[0, 1])
+
+    with app.app_context():
+        chunk_size = 2
+        returned_samples = get_cherrypicked_samples(root_sample_ids, plate_barcodes, chunk_size)
+        pd.testing.assert_frame_equal(expected, returned_samples)
+
+
+def test_get_cherrypicked_samples_sentinel_and_beckman(app, freezer):
```
Comment (andrewsparkes) [inline tests/helpers/test_reports_helpers.py:237]: can you add comments describing the scenario for this test? why is MCM001 in both sentinel and beckman dataframes?

File: setup_test_db.py [hunk 6]
```diff
@@ -6,29 +6,29 @@
 """Drop and recreate required tables."""
 print("Initialising the test MySQL MLWH database")
 
-sql_engine = create_mysql_connection_engine(config.WAREHOUSES_RW_CONN_STRING, config.ML_WH_DB)
+sql_engine = create_mysql_connection_engine(config.WAREHOUSES_RW_CONN_STRING)
 
-create_db = """
-CREATE DATABASE IF NOT EXISTS `unified_warehouse_test` /*!40100 DEFAULT CHARACTER SET latin1 */;
+create_db = f"""
+CREATE DATABASE IF NOT EXISTS `{config.ML_WH_DB}` /*!40100 DEFAULT CHARACTER SET latin1 */;
```
[the remark this answers is not in the document]
Comment (Chris-Friend) [reply setup_test_db.py:12]: This is a particularly good idea. Will revert

File: setup_test_db.py [hunk 7]
```diff
@@ -41,17 +41,20 @@
 `date_tested` datetime DEFAULT NULL COMMENT 'date_tested_string in date format',
 `source` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL COMMENT 'Lighthouse centre that the sample came from',
 `lab_id` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL COMMENT 'Id of the lab, within the Lighthouse centre',
+`lh_sample_uuid` varchar(36) DEFAULT NULL COMMENT 'Lighthouse sample unique identifier',
+`lh_source_plate_uuid` varchar(36) DEFAULT NULL COMMENT 'Lighthouse source plate unique identifier',
```
[the remark this answers is not in the document]
Comment (Chris-Friend) [reply setup_test_db.py:45]: Yes it is not used. But happy to leave it in for the same reason you point out

File: tests/helpers/test_reports_helpers.py [hunk 8]
```diff
@@ -74,35 +83,42 @@ def test_delete_reports(app, freezer):
...
+    ]
     samples = ["MCM001", "MCM002", "MCM003", "MCM004", "MCM005"]
     plate_barcodes = ["123", "456"]
 
     with app.app_context():
         with patch("sqlalchemy.create_engine", return_value=Mock()):
             with patch(
                 "pandas.read_sql",
-                return_value=expected,
+                side_effect=expected,
```
[the remark this answers is not in the document]
Comment (Chris-Friend) [reply tests/helpers/test_reports_helpers.py:103]: `side_effect` allows specifying a list of return values, such that the first value is returned on the first call to the method, the second value to the second call, and so on. This is about as flexible as pytest mocking gets w.r.t. different method return values

**Review by andrewsparkes — APPROVED** [review]

**Pull request merged**